### PR TITLE
Add package name in the docker image name

### LINKED ISSUE #713: These fields can be deprecated
https://github.com/viash-io/viash/blob/b91da037fa227d4bacc3d1b040a5cffee5a7f83d/src/main/scala/io/viash/engines/DockerEngine.scala#L54-L63

`target_registry`, `target_organisation`, `target_tag` should NOT be removed

### PR DESCRIPTION
**Pull request by Grifs**
Set DockerEngine package and organization default from package config

## Related issue(s)

Closes #713 

## Type of Change

- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] New functionality (non-breaking change which adds functionality)
- [ ] Major change (non-breaking change which modifies existing functionality)
- [x] Minor change (non-breaking change which does not modify existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] This change requires a documentation update

## Checklist

Requirements:

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc.
- [x] I have performed a self-review of my code by checking the "Changed Files" tab.
- [ ] My code follows the code style of this project.

Tests:

- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.

Documentation:

- [x] Proposed changes are described in the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/scala/io/viash/engines/DockerEngine.scala [hunk 1]
```diff
@@ -215,12 +223,17 @@ final case class DockerEngine(
...
+  def getTargetPackageNameWithFallback(config: Config): Option[String] = {
+    target_package.orElse(config.package_config.flatMap(_.name)).filter(_.nonEmpty)
+  }
+
   def getTargetIdentifier(config: Config): String = {
     val targetImageInfo = Docker.getImageInfo(
       config = Some(config),
       engineId = Some(id),
       registry = getTargetRegistryWithFallback(config),
       organization = target_organization,
```
Comment (Grifs) [inline src/main/scala/io/viash/engines/DockerEngine.scala:235]: Shouldn't this use a fallback value too? Currently this doesn't use a fallback value and it seems odd to me.

Comment (tverbeiren) [reply src/main/scala/io/viash/engines/DockerEngine.scala:235]: Not only that, I would suggest to add a placeholder/implicit organisation: `-` or `viash-hub` that can be overridden using `target_organisation` if need be.

Comment (Grifs) [reply src/main/scala/io/viash/engines/DockerEngine.scala:235]: The last few days I've been wondering ... would it make sense to add **another** config file, this time a global one.
Order of preference would then be `component config` > `package config` > `global config`.

Also, can someone explain me again why we have `organization` and `target_organization`, or `registry` and `target_registry`, ... ? I understand it's to be able to override the built container info, but why is there no fallback to the "non target_*" value?

**Comment by Grifs** [comment]
`.organization` is removed from the component config.

So far no additional fallback was added in case the `.organization` in package config is empty.

File: src/main/scala/io/viash/engines/DockerEngine.scala [hunk 2]
```diff
@@ -206,26 +215,26 @@ final case class DockerEngine(
...
-        |""".stripMargin           
+        |""".stripMargin
   }
 
-  def getTargetRegistryWithFallback(config: Config): Option[String] = {
-    target_registry.orElse(config.links.docker_registry)
-  }
+  def getTargetIdentifier(config: Config): DockerImageInfo = {
+
+    val targetRegistryWithFallback = target_registry.orElse(config.links.docker_registry)
```
Comment (rcannood) [inline src/main/scala/io/viash/engines/DockerEngine.scala:225]: If I look at this PR, would it then also not make sense to remove the component config > links > docker_registry and change this line to:

```suggestion
    val targetRegistryWithFallback = target_registry.orElse(config.flatMap(_.links.docker_registry)).filter(_.nonEmpty)
```

?

WDYT?

Comment (Grifs) [reply src/main/scala/io/viash/engines/DockerEngine.scala:225]: Surely it should be
```
val targetRegistryWithFallback = target_registry.orElse(config.links.docker_registry).filter(_.nonEmpty)
```
then. Config isn't an Option[...] here.
But sure, we can add this. The overriding of package config values is slightly different, but in the end, not that totally different.

File: src/main/scala/io/viash/engines/DockerEngine.scala [hunk 3]
```diff
@@ -194,6 +202,7 @@ final case class DockerEngine(
       name = Some(image),
       registry = registry,
       organization = organization,
+      `package` = None,
       tag = tag.map(_.toString),
       namespaceSeparator = namespace_separator
```
[the remark this answers is not in the document]
Comment (Grifs) [reply src/main/scala/io/viash/engines/DockerEngine.scala:207]: I removed it already now. The `engineId` was not being specified already, so there wasn't an actual reason why we should keep it here.
Once registry & organization are removed, it should clean up itself.
The `namespaceSeparator` was not used here, so I made it an option and removed it here.

File: src/main/scala/io/viash/engines/DockerEngine.scala (hunk 2 above)
Comment (rcannood) [reply src/main/scala/io/viash/engines/DockerEngine.scala]: :+1:

**Review by rcannood — CHANGES_REQUESTED** [review]
Would you like to add the changes suggested by comment #2 ?

**Comment by Grifs** [comment]
> Would you like to add the changes suggested by comment #2 ?

You mean this part?
```
    val targetRegistryWithFallback = target_registry.orElse(config.links.docker_registry).filter(_.nonEmpty)
```

If so, this is already present. If something else, could you please elaborate?

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,12 @@ TODO add summary
 
 * `resources_test`: This field is removed again from the `_viash.yaml` as it was decided to impliment this temporary functionality using the `info` field (PR #711).
 
+* `docker_engine`: Deprecate `registry`, `organization` and `tag fields` in the `docker_engine` (PR #712). Currently these are hardly ever used and instead the `image` field is used to specify the full image name.
```
Comment (rcannood) [inline CHANGELOG.md:9]:
```suggestion
* `docker_engine`: Deprecate `registry`, `organization` and `tag` fields in the `docker_engine` (PR #712). Currently these are hardly ever used and instead the `image` field is used to specify the full image name.
* ```

**Review by rcannood — APPROVED** [review]

**Comment by rcannood** [comment]
LGTM!

**Pull request merged**